### DISCUSSION TITLE
EON: A simpler configuration format

### DISCUSSION THREAD

**Discussion by brandonchinn178** [brandonchinn178/brandonchinn178.github.io#17] [Polls]

**Comment by dpwiz** [comment]
I'd like to have something like tags (EDN style or otherwise) to distinguish one-field-object-wrapper encoding of tagged objects from actual dicts with only one key.

Haskell value:

```haskell
pet = Cat { name = "Basil", length = cm 80 }
```

clojure/EDN:

```clojure
pet = #cat { :name "Basil" :length (#cm 80) }
```
yaml/wrapper:

```yaml
pet:
  cat:
     name: Basil
     length: { cm: 80 } # not that bad, but the type is changed to object
```
yaml/inline:

```yaml
pet:
  _type: cat # okay for dict values
  name: Basil
  length: { _type: cm, _value: 80 } # the others are getting the ugly
```

**Reply by brandonchinn178** [reply]
The problem with tags is that it's just unnecessary complication. Why not just encode it however (as an object `{ cm: 20 }` or a string `"20cm"`), and have the application parse it appropriately? The config shouldnt encode language-specific features (e.g. the difference between a "proper" object or a "wrapper" object), just an easy way to get data into the app

**Reply by dpwiz** [reply]
1) Element type may be changed: `"20cm"` was a number, and now it's a string. Or object.
2) Element type may *not* be changed: `{name: "Basil"}` and `{cat: {name: "Basil"}}` are both objects, which can trip up parsers.
3) Lack of native tagging promote ad-hoc solutions for parsring/tagging, sticking with unitless values, and/or peppering config files with comments.

**Reply by brandonchinn178** [reply]
First of all,

> `"20cm"` was a number, and now it's a string.

The fundamental distinction with EON is that _everything is a string_. It only carries semantic meaning (e.g. "20" is a number) if the application gives it meaning. This prevents, for example, the name of a song "20" from being implicitly cast as a number, and the application needing to remember to stringify numbers (or preventing the configurer from needing to remember to wrap it in quotes).The design goal here was that the configurer shouldnt have control over the types of things; the application should.

I'm not sure what your second example is trying to convey. If that object is at some key `foo`, the application already knows what _thing_ it expects at that key, and that _thing_ should know whether it's an object that should contain a `name` key or a `cat` key.

The third point reminds me a lot of YAML's tag feature, which 1) I never really liked, and 2) I can't think of any good example for, especially for EON, where the driving principles are: everything is a string, and the application drives the semantics.

The "ad-hoc solution for tagging" is a bit compelling, where perhaps EON could define

```
a: #foo
  ...
```
as syntax sugar for
```
a:
  tag: foo
  content:
    ...
```
requiring dedicated support in parsers (like we define for null/bools/datetimes). But personally, I find this too magical and unnecessary for what should be a very simple format. But, hey, if this ever gets off the ground, I'd be happy to gather feedback on this type of thing

**Reply by dpwiz** [reply]
> The fundamental distinction with EON is that everything is a string.

Ah, I see. Then parsing is of course is the way to go.

**Comment by dpwiz** [comment]
Also, code blocks for multiline text inclusions.

YAML has them. But then it ties into the tagged values problem:

```yaml
code:
  plaintext: >
    What is code?
    Baby don't hurt me
    Don't hurt me
    No more
```

~~~
code:
  ```plaintext
  Never gonna mark you up
  Never gonna let you down
  ```
~~~
